File: software/tca9555/tca9555.py
import smbus

class tca9555:
    def __init__(self, a2, a1, a0):
        self.I2CAddr = 32 | (a2 << 2 | a1 << 1 | a0)

    def setBit(self, a, bit, loc):
        mask = 1 << loc
        a &= ~mask
        if bit:
            a |= mask
        return a 

    def setPortDirection(self, portNum, pinNum, direction):
        if(~(pinNum >= 0 and pinNum <=7)):
            return 0            
        bus = smbus.SMBus(1)
        if(portNum == 0):
            data = bus.read_byte_data(self.I2CAddr, 6)
            data = self.setBit(data, direction, pinNum)
            bus.write_byte_data(self.I2CAddr, 6, data)
        elif(portNum == 1):
            data = bus.read_byte_data(self.I2CAddr, 7)
            data = self.setBit(data, direction, pinNum)
            bus.write_byte_data(self.I2CAddr, 7, data)
        bus.close()

    def setPortPolarity(self, portNum, pinNum, polarity):
        if(~(pinNum >= 0 and pinNum <=7)):
            return 0            
        bus = smbus.SMBus(1)
        if(portNum == 0):
            data = bus.read_byte_data(self.I2CAddr, 4)
            data = self.setBit(data, polarity, pinNum)
            bus.write_byte_data(self.I2CAddr, 4, data)
        elif(portNum == 1):
            data = bus.read_byte_data(self.I2CAddr, 5)
            data = self.setBit(data, polarity, pinNum)
            bus.write_byte_data(self.I2CAddr, 5, data)
        bus.close()
        
    def setOutputStates(self, portNum, bit):
        if(~(pinNum >= 0 and pinNum <=7)):
            return 0            
        bus = smbus.SMBus(1)
        if(portNum == 0):
            data = bus.read_byte_data(self.I2CAddr, 2)
            data = self.setBit(data, bit, pinNum)
            bus.write_byte_data(self.I2CAddr, 2, data)
        elif(portNum == 1):
            data = bus.read_byte_data(self.I2CAddr, 3)
            data = self.setBit(data, bit, pinNum)
            bus.write_byte_data(self.I2CAddr, 3, data)
        bus.close()

    def getInputStates(self, portNum):
        bus = smbus.SMBus(1)
        data =  bus.read_byte_data(self.I2CAddr, portNum)
        bus.close()
        return data

if __name__ == "__main__":
    tca = tca9555(0,0,0)
    data = 8
    pn = 7
    tca.setPortDirection(1,3,0)
    temp = tca.getInputStates(7)
    print(temp)

File: software/src/tca9555.cpp
#include <iostream>
#include <errno.h>
#include <wiringPiI2C.h>
#include "tca9555.h"

using namespace std;

TCA9555::TCA9555(int a)
{
    addr = a;
    fd = wiringPiI2CSetup(addr);
}

int TCA9555::setBit(int num, int loc, int x)
{
    return num ^= (-x ^ num) & (1UL << loc);
}

int TCA9555::setPortDirection(int portNum, int pinNum, int dir)
{
    // check inputs
    if(pinNum < 0 || pinNum > 7)
    {
        return -1;
    }
    else if(portNum != 1 || portNum != 0)
    {
        return -1;
    }

    // set register 
    int reg = 0x06;
    if(pinNum == 1)
    {
        reg = 0x07;
    }

    int data = getStates(reg);
    data = setBit(data, pinNum, dir);
    int result = wiringPiI2CWriteReg16(fd, reg, data);
    
    if(result == -1)
    {
        cout << "Error. Ernno is: " << errno << endl;
    }

    return 0;
}

int TCA9555::setPortPolarity(int portNum, int pinNum, int polarity)
{
    // check inputs
    if(pinNum < 0 || pinNum > 7)
    {
        return -1;
    }
    else if(portNum != 1 || portNum != 0)
    {
        return -1;
    }

    // set register 
    int reg = 0x04;
    if(pinNum == 1)
    {
        reg = 0x05;
    }

    int data = getStates(reg);
    data = setBit(data, pinNum, polarity);
    int result = wiringPiI2CWriteReg16(fd, reg, data);
    
    if(result == -1)
    {
        cout << "Error. Ernno is: " << errno << endl;
    }

    return 0;
}

int TCA9555::setOutputStates(int portNum, int pinNum, int bit)
{
    // check inputs
    if(pinNum < 0 || pinNum > 7)
    {
        return -1;
    }
    else if(portNum != 1 || portNum != 0)
    {
        return -1;
    }

    // set register 
    int reg = 0x02;
    if(pinNum == 1)
    {
        reg = 0x03;
    }

    int data = getStates(reg);
    data = setBit(data, pinNum, bit);
    int result = wiringPiI2CWriteReg16(fd, reg, data);
    
    if(result == -1)
    {
        cout << "Error. Ernno is: " << errno << endl;
    }

    return 0;
}

int TCA9555::getStates(int reg)
{
    // Add input check later
    return wiringPiI2CReadReg8(fd, reg); 
}

int main(void)
{
    return 0;
}
